indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java

```diff
@@ -91,151 +97,330 @@ public WorkerTaskMonitor(
-                taskStatus = taskStatusAlreadySubmitted.get();
-              } else {
-                log.info("Affirmative. Running task [%s]", task.getId());
-                workerCuratorCoordinator.unannounceTask(task.getId());
-                taskStatus = taskRunner.run(task).get();
-              }
-            }
-            catch (InterruptedException e) {
-              log.debug(e, "Interrupted while running task[%s], exiting.", task.getId());
-              return;
-            }
-            catch (Exception e) {
-              log.makeAlert(e, "I can't build there. Failed to run task")
-                 .addData("task", task.getId())
-                 .emit();
-              taskStatus = TaskStatus.failure(task.getId());
-            }
-            finally {
-              running.remove(task);
-            }
-
-            taskStatus = taskStatus.withDuration(System.currentTimeMillis() - startTime);
-
-            try {
-              workerCuratorCoordinator.updateAnnouncement(TaskAnnouncement.create(task, taskStatus));
-              log.info(
-                  "Job's finished. Completed [%s] with status [%s]",
-                  task.getId(),
-                  taskStatus.getStatusCode()
-              );
-            }
-            catch (Exception e) {
-              log.makeAlert(e, "Failed to update task status")
-                 .addData("task", task.getId())
-                 .emit();
+              notices.add(new RunNotice(task));
             }
           }
         }
     );
   }
 
-  private boolean isTaskRunning(final String taskId)
+  private void registerLocationListener()
   {
-    for (final Task runningTask : running) {
-      if (runningTask.getId().equals(taskId)) {
-        return true;
-      }
-    }
+    taskRunner.registerListener(
+        new TaskRunnerListener()
+        {
+          @Override
+          public void locationChanged(final String taskId, final TaskLocation newLocation)
+          {
+            notices.add(new LocationNotice(taskId, newLocation));
+          }
+        },
+        MoreExecutors.sameThreadExecutor()
+    );
+  }
+
+  private void addRunningTask(final Task task, final ListenableFuture<TaskStatus> future)
+  {
+    running.put(task.getId(), new TaskDetails(task, future));
+    Futures.addCallback(
+        future,
+        new FutureCallback<TaskStatus>()
+        {
+          @Override
+          public void onSuccess(TaskStatus result)
+          {
+            notices.add(new StatusNotice(task, result));
+          }
 
-    return false;
+          @Override
+          public void onFailure(Throwable t)
+          {
+            notices.add(new StatusNotice(task, TaskStatus.failure(task.getId())));
+          }
+        }
+    );
   }
 
   @LifecycleStop
   public void stop()
   {
     try {
-      pathChildrenCache.close();
       exec.shutdownNow();
+      pathChildrenCache.close();
```

Comment:
ah, no good reason. It was important in a previous edition of this changeset but doesn't really matter anymore. will put them back.